### ISSUE TITLE
Description

### ISSUE THREAD

**Issue by woowonjin** [woowonjin/Sarcasm_Detection#1]
### Sarcasm('비꼬기') 이란

- 비꼬기

  ['비꼬기'의 담화 구조와 운율적 특징에 관한 연구](http://www.riss.kr/search/detail/DetailView.do?p_mat_type=be54d9b8bc7cdb09&control_no=4848b4f11ab04777ffe0bdc3ef48d419&outLink=K)

  - '**맥락**에 의해 해석되어지는 화용적인 현상으로서 뚜렷한 대상에 대한 **부정적 평가**를 포함하는 평가 화행' 이라 정의

  - kreuz(1996) 에서 비꼬기를 돕는 5가지 단서를 분류

    1. 반대되는 사실

    2. 과장

    3. tag형 질문
    4. 직설적 표현
    5. 동작, 운율

    

- 문맥(맥락)이 없는 짧은 댓글(Contextless short sentence) 에서의 비꼬는 표현

  문맥이 포함되지 않는 짧은 댓글 등에서도 다음과 같은 요소들을 통해 '비꼬기' 를 판단할 수 있음. 

  - 작성자 또한 댓글에 동작,운율등이 적용되지 않는 점을 감안하여 비꼬는 표현에 충분한 단서를 포함시켜 작성.

  > ex. 아예 댓글로 논문을 쓰시지?

  

  1. 사전지식

     '댓글' 과 '논문' 의 scale 차이 등을 사전지식으로 가지고 있기에 표현상에서 모순이 발생함을 파악 가능

     - model에 이러한 배경지식을 어떻게 학습시킬 것인지가 관건
     - Pre-trained word embedding 으로 해결 가능할지 미지수

     

  2. 과장

     '댓글'에서 '논문' 으로 확장되는 scale의 과장이 포함됨.

     

  3. 직설적 표현

     '아예' , '쓰시지?' 등의 직설적이고 공격적인 표현을 통해 파악 가능

### Sarcasm Detection 관련 논문

[해외 논문들 모음][https://paperswithcode.com/task/sarcasm-detection]

[CASCADE(컨텐츠와 컨텍스트 중심의 하이브리드적 접근방법 모델)][https://arxiv.org/pdf/1805.06413v1.pdf ]

국내 논문 - [스포츠 기사 댓글에서 수사의문문으로 표현된 풍자 표현의 자동인식](http://www.dbpia.co.kr/journal/articleDetail?nodeId=NODE09300944)

- 수사 의문문 ''~하냐'' 에 집중

  > Ex.
  >
  > 니가 국가대표냐?
  >
  > 그 팀한테 1:0으로 이긴게 자랑이냐?

- 다음과 같이 클린봇 Demo 를 통해 `평판(긍/부정)` 과 `유쾌/불쾌` 등을 이용하여 `수사의문문` 관련 데이터를 수집할 수 있을 것으로 예상됨.(불쾌 + 긍/부정 없음)

### 데이터 수집

 `불쾌` expection 데이터 사용

현재 NEG / n / COMP 데이터 중 `불쾌` 값이 높고 `부정` 이 아닌 케이스

영화 리뷰 데이터에서 리뷰는 `부정` 이나 `Only PN model` 이 `긍정` 으로 예측한 데이터

비꼬는 표현이 들어간 기사나 제목등을 기준으로 제플린에서 데이터 크롤링(sarcasm이 표현된 제목일 수록 댓글에 sarcarsm이 포함될 가능성이 높다고 예상)

`긍/부정 평판` 에 사용된 25000여건의 데이터중, `불쾌` 값이 0.6 이상이고 `평가` label 이 n 인 데이터 17000 여건에 대해 labeling 진행  

### 모델 설계 계획

- ELMo 로 기존데이터 pretraining

- sarcasm 데이터에 대해 lr 감소시켜가며 fine-tuning

#### 1: 반어법/비꼬기 가 포함되어 있다.
 **부정 표현과 비꼬는 표현이 같이 존재하는 문장의 경우, 1로 태그**

- 부정적 의도의 수사 의문문

  > 환경단체 똘마니들은 뭐하냐??
  >
  > 술먹고기사썼나

  

- 부정적인 표현등을 불쾌한 감정표현 없이 사용하는 경우

  > 공산당이 먼저다. 문좨인입니다 훠훠훠

 

- 긍정 표현 + 부정 접속부사 (+ 부정표현)

  > 하차 안 해도 된다 어차피 그 드라마 안 보니까

- 모순적이고 과장된 표현

  > 손잡이에 다이아몬드 서너개 박고... 100억 짜리 냉장고 해라
  >
  > 왜 2100년에 한다하지?

  

- 창의적 표현이 포함된 비판 ( 언어유희 등 )

  > 왜국보수들 투표는 꼭 하세요
  >
  > 시청자들 뚜껑 열린 결말
  >
  > 미세먼지x 문세먼지o

- 돌려말하는 비하/부정 표현(부정적 평가 표현이 직접적이지 않은 경우)

  > 이런거는 의대생 기본 상식인데 공부못하는애들이 주로 가는 예방의학과 전문가들은 잘 모르나 봄.

- 인용/흉내 등을 통한 풍자(비웃음) 

  > ㅋㅋㅋㅋ 걸리게 생기니까 '기부할거에욧!!!'
  >
  > 훠훠훠,, 쩝쩝쩝

- '니,느그' 등 2인칭을 포함하는 비하의도의 명령문 표현

  > 느그 부모님들 돈 합쳐서 도와달라해라
  >
  > 답답하면 니들이 뛰던가

- 자주 사용되는 비꼬는 표현들도 포함

  > 우덜식 민주주의
  >
  > ???: 공부 좀 더 했다고 삼성전자가 임금 2배 더 받는 건 불공평
  >
  > 이게 나라냐
  >
  > 가지가지 한다, 참 대단하다 등
  >
  > 역시나 / ~ 클라쓰 ( 비꼬는 표현에서 자주 사용되는 긍정 과장 표현 )
  

#### e: 판단 불가능

아래의 두가지 경우는 e로 태깅한 뒤, training dataset에서 배제하여 모델의 통계적 판단에 맞깁니다. 

- 의미를 전혀 이해할 수 없는 문장

  

- 비꼬는 표현인 듯 하지만, 명확하지 않은 표현으로 context가 필요해 보이는 경우

### 데이터 평가 방식

감성분석에서 모델이 `긍정/중립`  이라고 판단한 데이터 중 실제로 `부정적인 비꼬기 ` 가 들어가있는 데이터를 추출하는 데에 의미가 있음 => **precision** 이 중요

ROC 커브를 보고 precision 을 극대화 하는 threshold 를 결정해볼 것

### 데이터 추가 확보

Zeppelin 에서 부정적 댓글을 포함할 것으로 예상되는 keyword 를 포함하는 title 의 댓글 목록을 crawling (9000여건 추가 태깅)

### 제목(title) 포함하여 크롤링

- Title 을 포함하여 데이터를 확보하였는데, 입력에 title 과 contents를 모두 주어 contextless 한 문제점을 해결할 수도 있을거라 판단 

- contents 를 부정적인 것으로 고정하고, title 이 포함된 것이 판단을 쉽게하여 태깅하는 데에 도움을 주었지만, 본질적으로 추구하던 contextless 한 조건을 무시할 수도 있어 주의가 필요할 것으로 생각됨.

  

`긍/부정 - 평가` 모델에서 prediction 하여 `부정` 수치가 낮게 나온 데이터들 중심으로 태깅(약 2000건의 sarcasm label data 추가 확보)

총 약 25000건 데이터 확보

#### Data Augmentation

Noise Augmentation

- 오타생성

- g2pK 라이브러리 이용

  

Keyboard Augmentation

- Sub Character Deletion
- Jaso Seperation
- Swap Position

Back translation

데이터 3000여개 추가, 성능 약간 증가

Oversampling 한 데이터로 적용한 모델과 비교시 큰 accuracy/f1_score 차이를 보이지는 않음. 

#### ELMo 사용

- C3DL 권한 회수로 로컬환경에서 CPU 사용하여 Model fork 하여 사용
- 자소&음절 Embedding / ELMo 모델

#### GAN 활용

- 댓글등의 Short Sentence 등에서 GAN이 동작 효율이 떨어짐(데이터 수집이 간단하고, 변형이 많아 GAN으로 만들어낼 수 있는 데이터의 정확도가 보장되지 않음.)

- 추후 데이터를 보고 진행 예정

### Out Of Distribution(OOD) 활용 

비꼬는 문장의 특성상 기존의 단어들을 변형하거나 창의적으로 만들어내는 경우가 자주 발견됨

틀에 박힌 비꼬는 표현의 판단은 모델이 쉽게 파악할 수 있지만, 이러한 창의성이 포함된 비꼬는 표현들은

자주 등장하지 않고, 모델의 판단이 어려워 보임.

Out of Distribution Dectection 을 활용하는 것으로 이러한 문제를 따로 풀어볼 수 있을 것으로 예상됨.

### 계획 및 일정

논문 작성 후 국내 학술지에 계획
